src/librustc_typeck/collect.rs

```diff
@@ -50,9 +49,65 @@ use rustc::hir::GenericParamKind;
 use rustc::hir::{self, CodegenFnAttrFlags, CodegenFnAttrs, Unsafety};
 
 use std::iter;
+use std::ops::Range;
 
 struct OnlySelfBounds(bool);
 
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub enum LintedBoundVar<'tcx> {
+    DefId(DefId),
+    SelfTy,
+    Ty(Binder<Ty<'tcx>>),
+    Region(Binder<ty::Region<'tcx>>),
+}
+
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub enum LintedBoundValue<'tcx> {
+    DefId(DefId),
+    Ty(Binder<Ty<'tcx>>),
+    Region(Binder<ty::Region<'tcx>>),
+}
+
+pub fn lint_duplicate_bounds<'a, 'gcx, 'tcx>(
+    tcx: TyCtxt<'a, 'gcx, 'tcx>,
+    node_id: ast::NodeId,
+    bounds: impl IntoIterator<Item = (LintedBoundVar<'tcx>, LintedBoundValue<'tcx>, Span)>,
+) {
+    let mut bounds: Vec<_> = bounds.into_iter().collect();
+    bounds.sort_unstable();
+
+    let emit_lint = |range: Range<usize>| {
+        let bound_name = tcx.sess.source_map()
+                            .span_to_snippet(bounds[range.start].2).unwrap();
+        let mut err = tcx.struct_span_lint_node(
+            lint::builtin::DUPLICATE_BOUNDS,
+            node_id,
+            bounds[range.clone()].iter().map(|(_, _, sp)| *sp).collect::<Vec<_>>(),
+            &format!("duplicate bound `{}` found",
+                     bound_name));
+        debug!("zzz: 1: {:?} / {:?}", bounds[range.start].0, bounds[range.start].1);
+        err.span_label(bounds[range.start].2, "first use of bound");
+        for i in (range.start + 1)..range.end {
+            debug!("zzz: 2: {:?} / {:?}", bounds[i].0, bounds[i].1);
+            err.span_label(bounds[i].2, "subsequent use of bound");
+        }
+        err.emit();
+    };
+
+    let mut seq_start = 0;
+    for i in 1..bounds.len() {
+        if (&bounds[i].0, &bounds[i].1) != (&bounds[i - 1].0, &bounds[i - 1].1) {
```

Comment:
You can use `tuple_windows` from [itertools](https://docs.rs/itertools/0.8.0/itertools/trait.Itertools.html#method.tuple_windows) if available to express this more nicely.

Reply:
If only... introducing dependencies on itertools in the compiler is a big no no.

Reply:
Bummer... I would have expected it to already be used somewhere... :/